### PR TITLE
Adds Datum Tagging for admins

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
As an admin who likes to mess around with VV and do little events, and as someone who likes bouncing between observing different points of interest, I often get annoyed that my best option for keeping track of them is to just keep their VV menus open but minimized in case I want it later. But no more! This PR adds the Tag Datum verb and Tag Menu!

This PR adds the ability to "Tag" datums so you can pull them up in a menu after the fact, pictured below. You can toggle an atom's tag by **ctrl+middle clicking** or by using the right click context menu. You can also tag things from their VV dropdown menu, useful for abstract datums. To open the Tag menu, the default hotkey is **F9**, or you can use the View-Tags verb found in the Admin -> Game tab. 

[](https://i.imgur.com/wgKbRIK.png)

Some datum types show extra information about their status, like how you can see the corgi's health total, and the human's 4 damage types. Subsystems will show their stat panel entry, though obviously the panel doesn't auto update.

[](https://i.imgur.com/NUeCc9F.png)

In addition, you can also use tagged datums in VV edits and proccall arguments in the same way you could with Mark Object, except this lets you select from multiple choices. The only reason I'm not getting right of Mark Object in this PR entirely is SDQL, which would require more attention to replace marking's functionality than I want to put into this PR.

## Why It's Good For The Game
Makes it easier for admins/coders to keep track of different datums/atoms around the game world.

## Changelog

:cl: Ryll/Shaps
admin: Added Datum Tagging, a new admin tool for quickly marking and referring back to various datums and atoms. You can tag atoms by ctrl+middle clicking them or using the right click context menu, or from their VV dropdown menu (especially for abstract datums that can't be clicked).
admin: The View-Tags menu can quickly be accessed by F9 default, or using the View-Tags button at the bottom of the Admin -> Game status panel section.
admin: Tagged datums can also be referenced for things like proccalls and VV var edits, similarly to Marked objects.
/:cl:

### CODE REVIEW

**Comment by Mothblocks** [comment]
doooope

**Comment by tralezab** [comment]
super dope + obligatory i wish you tgui'd it

File: code/modules/admin/tag.dm [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+ * Arguments:
+ * * target_datum - The datum you want to remove from the tagged_datums list. Can also be the specific weakref used in said list.
+ * * silent - If TRUE, won't print messages to the owner's chat
+ */
+/datum/admins/proc/remove_tagged_datum(datum/target_datum, silent=FALSE)
+	if(isnull(target_datum))
+		return
+
+	if(!istype(target_datum)) // for handling memory REF(X)'s to the target datum
+		target_datum = locate(target_datum)
```
Comment (LemonInTheDark) [inline code/modules/admin/tag.dm:43]: What would cause this case?

File: code/modules/admin/tag.dm [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+			to_chat(owner, span_warning("[target_datum] was not already tagged."))
+		return
+	else // we have to actually resolve each tagged weakref to see which one is ours
+		for(var/datum/weakref/iter_weakref as anything in tagged_datums)
+			if(iter_weakref.resolve() == target_datum)
+				LAZYREMOVE(tagged_datums, iter_weakref)
+				UnregisterSignal(target_datum, COMSIG_PARENT_QDELETING)
+				if(!silent)
+					to_chat(owner, span_notice("[target_datum] has been untagged."))
+				return
```
Comment (LemonInTheDark) [inline code/modules/admin/tag.dm:60]: You don't actually need this loop. Each datum will only ever have one weakref attached to it, so something like this would be valid.

I have a feeling that'd clean up the bit above, but it may become a bit meaningless if you decide to ditch weakrefs 
```suggestion
	else // we have to actually resolve each tagged weakref to see which one is ours
		var/datum/weakref/ref_to_us = WEAKREF(target_datum)
		if(!(ref_to_us in tagged_datums)
			//Failure case
			return  
		LAZYREMOVE(tagged_datums, ref_to_us)
		UnregisterSignal(target_datum, COMSIG_PARENT_QDELETING)
		if(!silent)
			to_chat(owner, span_notice("[target_datum] has been untagged."))
		return
```

File: code/modules/admin/tag.dm [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+/datum/admins/proc/add_tagged_datum(datum/target_datum)
+	var/datum/weakref/new_weakref = WEAKREF(target_datum)
+
+	if(LAZYFIND(tagged_datums, new_weakref))
+		new_weakref = null // needed?
+		to_chat(owner, span_warning("[target_datum] is already tagged!"))
+		return
+
+	LAZYADD(tagged_datums, new_weakref)
+	RegisterSignal(target_datum, COMSIG_PARENT_QDELETING, .proc/handle_tagged_del)
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/modules/admin/tag.dm:18]: (If you do change to just refs do remember to clear the tagged_datums list on admin holder qdel, since the signals will just cease to exist but the list may continue on

File: code/modules/admin/tag.dm [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+	for(var/datum/weakref/iter_weakref as anything in tagged_datums)
+		if(!iter_weakref.resolve())
+			LAZYREMOVE(tagged_datums, iter_weakref)
+			cleared_nulls++
+
+	if(cleared_nulls)
+		to_chat(owner, span_notice("[cleared_nulls] nulls cleared."))
+	else
+		to_chat(owner, span_warning("No nulls found."))
```
Comment (LemonInTheDark) [inline code/modules/admin/tag.dm:77]: Why is this needed if you're using signals already? ui updates and such?

File: code/modules/admin/topic.dm [hunk 5]
```diff
@@ -2051,3 +2051,21 @@
...
 			return
 		GLOB.interviews.ui_interact(usr)
+
+	else if(href_list["del_tag"])
+		if(!check_rights(R_ADMIN))
+			return
+		var/datum/datum_to_remove = locate(href_list["del_tag"])
+		if(!datum_to_remove)
+			return
+		return remove_tagged_datum(datum_to_remove)
```
Comment (LemonInTheDark) [inline code/modules/admin/topic.dm:2061]: feels wrong but hell it's an admin menu they could do a lot more harm then just this

File: code/modules/admin/tag.dm [hunk 6]
```diff
@@ -0,0 +1,128 @@
+/**
+ * Creates a weakref of the target datum and inserts it into [/datum/admins/var/tagged_datums], for later reference.
+ *
+ * Note that if the original target is deleted and another datum is put in the same memory address, tthis
+ *
```
Comment (LemonInTheDark) [inline code/modules/admin/tag.dm:5]: This feels like it was forgotten as a part of a rewrite

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
Worth recognizing this is something close to marked datums++, but I suppose that being more targeted is somewhat useful. Using tagged[number] in sdql as syntax would be weird and all

**Comment by Razharas** [comment]
Does this not overlap with marked objects?

**Comment by Ryll-Ryll** [comment]
You can only have one marked object at a time

**Comment by LemonInTheDark** [comment]
So it's effectively marked objects++, right?

**Comment by Razharas** [comment]
if this is marked objects++ then you might want to remove amrked objects and replace them with tags completely
its not good to have options of nearly same usability but done differently for historical reasons

**Comment by Ryll-Ryll** [comment]
Forgot to upload my commit changing it from weakrefs to normal refs

As for replacing marked object with tagged datums, as potato mentioned in a comment above, I wouldn't really know how to implement that alongside SDQL at this moment. I can look into having tagged datums supplant marked objects in the near future, but it's out of scope for this PR

**Comment by Mothblocks** [comment]
There's a warning lint but I'll look at this soon

File: code/modules/admin/view_variables/tag_datum.dm [hunk 7]
```diff
@@ -0,0 +1,9 @@
+/client/proc/tag_datum(datum/D)
+	if(!holder || !D)
+		return
+	holder.add_tagged_datum(D)
+
+/client/proc/tag_datum_mapview(datum/D as mob|obj|turf|area in view(view))
+	set category = "Debug"
+	set name = "Tag Object"
+	tag_datum(D)
```
Comment (Mothblocks) [inline code/modules/admin/view_variables/tag_datum.dm:9]:
```suggestion
/client/proc/tag_datum(datum/datum)
	if(!holder || QDELETED(datum))
		return
	holder.add_tagged_datum(datum)

/client/proc/tag_datum_mapview(datum/datum as mob|obj|turf|area in view(view))
	set category = "Debug"
	set name = "Tag Object"
	tag_datum(datum)
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
While we still have marked objects, can you add a way to mark an object from this menu?

This seems like it'd be very useful when I'm doing big advanced proc calls, and using the "marked object" entry.

**Comment by Ryll-Ryll** [comment]
[](https://i.imgur.com/IUKHvbt.png)

Updated with more context info and fulfilling a few requests
